crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs

```diff
@@ -0,0 +1,171 @@
+    let kind = &checker.ctx.scope().kind;
+    if let ScopeKind::Function(FunctionDef {
+        name: caller_name, ..
+    }) = *kind
+    {
+        let Located { node, .. } = func;
+        if let Some(scope) = checker.ctx.parent_scope() {
+            match scope {
+                Scope {
+                    // caller is defined in class
+                    kind:
+                        ScopeKind::Class(
+                            ClassDef {
+                                name: class_name, ..
+                            },
+                            ..,
+                        ),
+                    ..
+                } => {
+                    match node {
+                        // callee belongs to same class as caller
+                        ExprKind::Attribute {
+                            attr: callee_name, ..
+                        } => {
+                            return func_defined_above(
+                                checker,
+                                func,
+                                scope,
+                                callee_name,
+                                caller_name,
+                            );
+                        }
+                        // callee belongs to module, scope switches to module and class becomes caller
+                        ExprKind::Name {
+                            id: callee_name, ..
+                        } => {
+                            if let Some(scope) = checker.ctx.grandparent_scope() {
+                                return func_defined_above(
+                                    checker,
+                                    func,
+                                    scope,
+                                    callee_name,
+                                    class_name,
+                                );
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+                Scope {
+                    // caller is defined in module
+                    kind: ScopeKind::Module,
+                    ..
+                } => {
+                    // callee belongs to module
+                    if let ExprKind::Name {
+                        id: callee_name, ..
+                    } = node
+                    {
+                        return func_defined_above(checker, func, scope, callee_name, caller_name);
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+    None
+}
+
+// check if callee function that is called from caller is defined above caller
+fn func_defined_above(
+    checker: &Checker,
+    func: &Expr,
+    scope: &Scope,
+    callee_name: &String,
+    caller_name: &str,
+) -> Option<Diagnostic> {
+    if let (Some(callee_binding_idx), Some(caller_binding_idx)) =
+        (scope.get(&callee_name[..]), scope.get(caller_name))
+    {
+        if caller_binding_idx > callee_binding_idx {
```

Comment:
Hmm, if I understand what's going on here, I consider this an implementation detail. It's not guaranteed that these bindings are increasing as we parse through the program. Can we instead look at the `source` field on `Binding`, and compare based on location?

Reply:
Fixed